### PR TITLE
Profile page 1

### PR DESCRIPTION
**Pull request by Iamogeee**
### Summary of Changes

Summary of Changes:

**Added Profile Route:**

Implemented a new route for the user profile page, enabling users to access and manage their profiles.

**Updated Schema:**

Made necessary updates to the database schema to accommodate the new profile features, ensuring proper data storage and retrieval.

**Installed Multer:**

Added Multer to the project dependencies to enable file upload handling.

**Created New Endpoints and Set Up Multer for Uploads:**

Developed new API endpoints to support profile-related functionalities, including profile photo uploads.
Integrated Multer, a middleware for handling multipart/form-data, to manage file uploads efficiently.

### CODE REVIEW

File: backend/index.js [hunk 1]
```diff
@@ -123,10 +150,16 @@ app.get("/api/businesses", async (req, res) => {
 app.get("/api/search", async (req, res) => {
   const { query, userId } = req.query;
 
+  const cacheKey = `search:${userId}:${query}`;
+  const cachedResults = await redisCache.get(cacheKey);
```
Comment (sinap-fb) [inline backend/index.js:154]: awesome, this caching is so succinct!

File: backend/index.js [hunk 2]
```diff
@@ -177,6 +230,57 @@ app.get("/api/user", authenticateToken, async (req, res) => {
...
+      .status(500)
+      .json({ error: "Failed to fetch user data", details: error.message });
+  }
+});
+
+// Update user's profile
+app.put(
+  "/api/user/profile",
+  authenticateToken,
+  upload.single("profilePicture"),
```
Comment (sinap-fb) [inline backend/index.js]: would you mind adding a comment for what this line does? (i'm unfamiliar with this library)

File: backend/invertedIndex.js [hunk 3]
```diff
@@ -2,13 +2,25 @@ const { PrismaClient } = require("@prisma/client");
...
-    .toLowerCase()
-    .replace(/[^a-z0-9\s]/g, "")
-    .split(/\s+/);
+  // Convert text to lowercase and remove non-alphanumeric characters (except spaces and underscores)
+  const normalizedText = text.toLowerCase().replace(/[^a-z0-9\s_]/g, "");
+
+  // Split text into words based on spaces
+  let words = normalizedText.split(/\s+/);
+
+  // Split words based on underscores and flatten the resulting arrays
```
Comment (sinap-fb) [inline backend/invertedIndex.js:11]: these sort of comments are very helpful!

File: backend/redisCache.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
+const redis = require("redis");
+const { promisify } = require("util");
+
+class RedisCache {
+  constructor(ttl = 3600) {
+    // Create a Redis client with built-in Promise support
+    this.client = redis.createClient({
+      legacyMode: true, // This enables legacy mode which includes the callback API
```
Comment (sinap-fb) [inline backend/redisCache.js:8]: i'm not familiar with Redis - why do we want the callback API when it seems like we're using promises for everything?

File: backend/invertedIndex.js [hunk 5]
```diff
@@ -18,21 +30,42 @@ const buildInvertedIndex = async () => {
...
-module.exports = { buildInvertedIndex };
+// Global variable to store the inverted index
+let invertedIndex;
+
+const initializeInvertedIndex = async () => {
+  invertedIndex = await buildInvertedIndex();
+};
+
+// Initialize the inverted index when the module is loaded
+initializeInvertedIndex();
```
Comment (sinap-fb) [inline backend/invertedIndex.js:64]: hm, why do we call `buildInvertedIndex()` when the module is loaded? It seems like we call the same function in `searchBusinesses()`.  Does the index need to be built once at startup, or does it need to happen on each request that accesses the index (such that the index is up-to-date)?

File: backend/search.js [hunk 6]
```diff
@@ -1,28 +1,36 @@
...
   });
 
-  return results;
+  // Convert the results map to an array and sort by match count in descending order
+  const resultsArray = Array.from(resultsMap.values())
+    .sort((a, b) => b.matchCount - a.matchCount)
+    .map((entry) => entry.business);
+
+  // Return the sorted array of businesses
+  return resultsArray;
```
Comment (sinap-fb) [inline backend/search.js:33]: the comments in this function are fantastic! This is really great for readability.

File: frontend/src/Components/BusinessDetailsSkeleton/BusinessDetailsSkeleton.css [hunk 7]
```diff
@@ -0,0 +1,80 @@
+.business-details-skeleton {
+  padding: 0px 0 20px 0;
+}
+
+.skeleton-Header {
```
Comment (sinap-fb) [inline frontend/src/Components/BusinessDetailsSkeleton/BusinessDetailsSkeleton.css]: super nitpick, but I'd try to keep the naming convention consistent
```suggestion
.skeleton-header {
```

File: frontend/src/Components/BusinessDetailsSkeleton/BusinessDetailsSkeleton.jsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+          <div className="skeleton-hour shimmer"></div>
+          <div className="skeleton-hour shimmer"></div>
+          <div className="skeleton-hour shimmer"></div>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+export default BusinessDetailsSkeleton;
```
Comment (sinap-fb) [inline frontend/src/Components/BusinessDetailsSkeleton/BusinessDetailsSkeleton.jsx:36]: would love to see how this looks in action! (maybe you have a video in a later PR?)

File: frontend/src/Components/ProtectedRoutes/ProtectedRoutes.jsx [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import React from "react";
+import { Route, Navigate } from "react-router-dom";
+
+const ProtectedRoute = ({ element: Component, ...rest }) => {
+  const isAuthenticated =
+    localStorage.getItem("user") && localStorage.getItem("token");
```
Comment (sinap-fb) [inline frontend/src/Components/ProtectedRoutes/ProtectedRoutes.jsx:6]: maybe you plan on modifying this in a later PR, but this isn't a good authentication check.  Any user can put a dummy value in their local storage such that `isAuthenticated` becomes true.  Ideally we actually somehow validate that the token's value is a valid, non-expired token.

**Review by sinap-fb — COMMENTED** [review]
Nicely done. I think the code readability has improved a lot, especially with the descriptive comments

File: backend/index.js (hunk 2 above)
Comment (Iamogeee) [reply backend/index.js]: Done

File: backend/redisCache.js (hunk 4 above)
Comment (Iamogeee) [reply backend/redisCache.js:8]: I've actually switched from this approach to another in a later PR

File: backend/invertedIndex.js (hunk 5 above)
Comment (Iamogeee) [reply backend/invertedIndex.js:64]: We call buildInvertedIndex() when the module is loaded to ensure that the inverted index is built once at startup. However, in searchBusinesses(), we also call buildInvertedIndex() to ensure the index is up-to-date.

File: frontend/src/Components/BusinessDetailsSkeleton/BusinessDetailsSkeleton.css (hunk 7 above)
Comment (Iamogeee) [reply frontend/src/Components/BusinessDetailsSkeleton/BusinessDetailsSkeleton.css]: Alright

File: frontend/src/Components/ProtectedRoutes/ProtectedRoutes.jsx (hunk 9 above)
Comment (Iamogeee) [reply frontend/src/Components/ProtectedRoutes/ProtectedRoutes.jsx:6]: I would love some more clarity on this if you don't mind. And how would i would go about doing this

**Pull request merged**